rfcs/0048-rosetta-zkapps.md

```diff
@@ -0,0 +1,119 @@
+
+The changes needed for Rosetta to support these changes are outlined in the [0045 RFC](./0045-zkapp-balance-data-in-archive.md) under the "Changes to Rosetta" sub-heading.
+
+To summarize, for the `Account` endpoint, the SQL queries can be simplified to merely a pending and canonical query to the new `accounts_accessed` table; we no longer need to inspect individual transactions.
+
+For the `Block` endpoint, adjust the queries to not rely on the bits that are no longer present in the existing tables, and add a new SQL query to use the `accounts_accessed` table to pull the account-creation-fee parts.
+
+### Handling zkApp transactions
+
+zkApp Transactions will introduce many additions to our Rosetta implementation, but in an effort to keep the scope down, there are a few areas of Rosetta that will not support them.
+
+Rosetta's spec demands that it be able to track all balance changes that result from transactions on the network; however, we don't need to be able to construct all such balances nor do we need to track balance changes that only involve custom tokens. In order to support custom tokens built on top of Mina we would need to both be able to broadcast zkApp transactions and also keep track of token changes.
+
+#### Construction API
+
+As such, custom tokens are _out of scope_ and thus the Rosetta Construction flow will _not_ need to change. Instead, we will continue to support the existing legacy transactions.
+
+#### Data API
+
+The [Data API](https://www.rosetta-api.org/docs/data_api_introduction.html) will need to change to support zkApps.
+
+Most of the changes will be localized to supporting new zkApps related [Rosetta operations](https://www.rosetta-api.org/docs/models/Operation.html) and putting them in a new type of [Rosetta transaction](https://www.rosetta-api.org/docs/models/Transaction.html).
+
+These new transactions will be present in both the `/block` and `/mempool` endpoints.
+
+Note: GraphQL queries for zkApp transactions can be [generated programatically](https://github.com/MinaProtocol/mina/blob/develop/src/lib/mina_base/parties.ml#L1431) as they are quite large.
+
+**Operations for zkApps**
+
+[Operation types](https://github.com/MinaProtocol/mina/blob/35ed5e191af9cfa2709f567f6fe85d96dabfafef/src/lib/rosetta_lib/operation_types.ml) should be extended with new operations for all kinds of ways zkApp transactions can manipulate account balances (for each one-sided changes).
+
+Luckily this is made extremely clear by the shape of the zkApps transaction structure.
+
+The following intends to be an exhaustive list of those sorts of operations:
+
+1. `Zkapps_fee_payer_dec`
+2. `Zkapps_balance_change`
+
+Note that only balance changes that correspond with the default MINA token (`token_id` = 1) should be considered as changes here.
```

Comment:
`Token_id.t` has changed, it's no longer the integer 1, it's a field derived from 1. It shows up in the db as a string that you wouldn't guess is 1. Probably just shouldn't mention it.

Reply:
I checked and I think the `token_id` for MINA is hardcoded to be 1 still (unless we encode that 1 in some different way?)

It is important for us to check it because we need to ignore operations on tokens of other shapes.

Reply:
It's this: `Snark_params.Tick.Field.one`, which is different than the uint64 1 we had before. The printed representation is not `1`, it's a Bas58Check value.